### PR TITLE
forget password feature

### PR DESCRIPTION
**Pull request by 13jksingh**
Fixes https://github.com/PEC-CSS/acm-website-backend/issues/81

### CODE REVIEW

File: src/main/java/com/pecacm/backend/controllers/UserController.java [hunk 1]
```diff
@@ -143,6 +145,28 @@ public ResponseEntity<User> updateUser(@RequestBody User user) {
         return (updatedUser == null) ? ResponseEntity.badRequest().build() : ResponseEntity.ok(updatedUser);
     }
 
+    @GetMapping("/forget-password")
+    @PreAuthorize(Constants.HAS_ANY_ROLE)
+    public ResponseEntity<ForgetPasswordResponse> getForgetPasswordToken(String email){
+        User user = userService.getUserByEmail(email);
+        ForgetPasswordToken forgetPasswordToken = forgetPasswordService.getForgetPasswordToken(user);
+        return ResponseEntity.status(HttpStatus.OK).body(new ForgetPasswordResponse(forgetPasswordToken.getToken().toString()));
+    }
```
Comment (harasees-singh) [inline src/main/java/com/pecacm/backend/controllers/UserController.java:154]: aren't you sending back the verification token in response to a GET request with provided email ? this way anyone would be able to make a GET request with someone else's email and change their password.

**Review by harasees-singh — COMMENTED** [review]
requires changes

File: src/main/java/com/pecacm/backend/services/UserService.java [hunk 2]
```diff
@@ -201,4 +201,9 @@ public List<Transaction> getUserTransactions(String email, Integer offset, Integ
         transactionRepository.findByUser_EmailOrderByDateDesc(email, PageRequest.of(offset, pageSize)).forEach(transactions::add);
         return transactions;
     }
+    @Transactional
+    public void changePassword(User user, String newPassword) {
+        user.setPassword(newPassword);
+        userRepository.save(user);
+    }
```
Comment (harasees-singh) [inline src/main/java/com/pecacm/backend/services/UserService.java:208]: encode the password before saving it to db

File: src/main/java/com/pecacm/backend/repository/ForgetPasswordTokenRepository.java [hunk 3]
```diff
@@ -0,0 +1,11 @@
+package com.pecacm.backend.repository;
+
+import com.pecacm.backend.entities.ForgetPasswordToken;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.stereotype.Repository;
+
+import java.util.UUID;
+@Repository
+public interface ForgetPasswordTokenRepository extends JpaRepository<ForgetPasswordToken, UUID> {
+    ForgetPasswordToken findByToken(UUID token);
```
Comment (ken1000minus7) [inline src/main/java/com/pecacm/backend/repository/ForgetPasswordTokenRepository.java:10]:
```suggestion
    Optional<ForgetPasswordToken> findByToken(UUID token);
```

File: src/main/java/com/pecacm/backend/services/ForgetPasswordService.java [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    @Transactional
+    public ForgetPasswordToken getForgetPasswordToken(User user) {
+        return forgetPasswordTokenRepository.save(
+                ForgetPasswordToken.builder().user(user).build()
+        );
+    }
+
+    public ForgetPasswordToken findByToken(UUID token) {
+        return forgetPasswordTokenRepository.findByToken(token);
+    }
```
Comment (ken1000minus7) [inline src/main/java/com/pecacm/backend/services/ForgetPasswordService.java:30]: Throw 404 if token doesn't exist

**Review by ken1000minus7 — CHANGES_REQUESTED** [review]

**Comment by harasees-singh** [comment]
issue solved, closing PR

**Pull request closed without merging**